[PIP] Pipelines In Pipelines

jerop opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

This is an experimental project to provide support for `Pipelines` in
`Pipelines` to improve the composability and reusability of [Tekton Pipelines](https://github.com/tektoncd/pipeline). 
The functionality is provided by a controller that implements the 
`Custom Task` interface.

Its use cases include enabling defining and executing a set of `Tasks`
as a complete unit of execution and decoupling `Pipeline` branches
failures such that a `Task` failure in one branch does not stop
execution of unrelated `Pipeline` branches.

Pipelines in Pipelines is proposed in tektoncd/community#330
Issue is further described in tektoncd/pipeline#2134

Future work includes enabling passing in inputs through `Parameters`, 
producing outputs through `Results`, propagating timeouts, among others. 

/kind feature

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Commit messages includes a project tag in the subject - e.g. [OCI], [hub], [results], [task-loops]

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md) for more details._

/cc @afrittoli @bobcatfish @ImJasonH

afrittoli commented on pipelines-in-pipelines/config/100-namespace.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: tekton-pip-run
```
Any special reason for running this in a different namespace by default?

jerop replied:
we run them in a separate namespace to make it clear what was created by Pipelines in Pipelines controller, and make it easy to delete all its resources as needed

jerop replied:
if/when the feature is promoted from experimental to top level, we can switch to `tekton-pipelines` namespace at that time

jerop replied:
/cc @ImJasonH

afrittoli commented on pipelines-in-pipelines/config/config-logging.yaml:
```diff
@@ -0,0 +1,53 @@
+      "errorOutputPaths": ["stderr"],
+      "encoding": "json",
+      "encoderConfig": {
+        "timeKey": "",
```
NIT: I think it would be nice to have a timestamp key here so we can get timestamps in the logs.

jerop replied:
sounds good, added :)

afrittoli commented on pipelines-in-pipelines/go.mod:
```diff
@@ -0,0 +1,33 @@
+)
+
+// Knative deps
+replace (
+	contrib.go.opencensus.io/exporter/stackdriver => contrib.go.opencensus.io/exporter/stackdriver v0.12.9-0.20191108183826-59d068f8d8ff
+	github.com/Azure/azure-sdk-for-go => github.com/Azure/azure-sdk-for-go v38.2.0+incompatible
+	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.4.0+incompatible
+)
+
+// Pin k8s deps to v0.18.8
+replace (
+	k8s.io/api => k8s.io/api v0.18.8
+	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.18.8
+	k8s.io/apimachinery => k8s.io/apimachinery v0.18.8
+	k8s.io/client-go => k8s.io/client-go v0.18.8
+	k8s.io/code-generator => k8s.io/code-generator v0.18.8
+	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20200410145947-bcb3869e6f29
+)
```
Do we need these replace? 
It's not a problem if we do, just something that we'd need to maintain in future.

jerop replied:
yes yes, otherwise I hit some issues like this https://github.com/kubernetes/client-go/issues/741 

we use the same deps in cel custom tasks, so can rework them separately

afrittoli commented on pipelines-in-pipelines/pkg/reconciler/pip/piprun.go:
```diff
@@ -0,0 +1,258 @@
+		logger.Errorf("Run %s/%s is invalid because of %v", run.Namespace, run.Name, err)
+		run.Status.MarkRunFailed(ReasonRunFailedValidation,
+			"Run can't be run because it has an invalid spec - %v", err)
+		return nil
```
It would be nice to return a [permanent error](https://github.com/knative/eventing/blob/9b0599f33985020f445820ce6ee843d5ab701122/vendor/knative.dev/pkg/controller/controller.go#L425) instead.
The controller will not re-queue the key in case of permanent errors, but it will be aware about the queue (and an event will be emitted for it).

jerop replied:
great idea, changed it to return permanent error :)

afrittoli commented on pipelines-in-pipelines/pkg/reconciler/pip/piprun_test.go:
```diff
@@ -0,0 +1,381 @@
+	if run.Status.StartTime == nil {
+		t.Errorf("Expected Run start time to be set but it wasn't")
+	}
+	if expectedStatus == corev1.ConditionUnknown {
```
Nice :)

afrittoli commented on pipelines-in-pipelines/pkg/reconciler/pip/piprun_test.go:
```diff
@@ -0,0 +1,381 @@
+	}
+}
+
+func checkEvents(fr *record.FakeRecorder, testName string, wantEvents []string) error {
```
This function is getting copied in a few places, I wonder if we should export it from a test packages somewhere... but it's ok to copy it here :)

jerop replied:
agreed, will need to do that because it's used to check events in all the custom tasks controllers

afrittoli reviewed (commented):
Thanks for this 🎉 🙏 
It looks great, I added a couple of NITs / questions / comments, but nothing that cannot be addressed in a different patch.
/approve

afrittoli commented:
Perhaps we could merge https://github.com/tektoncd/experimental/pull/706 first so we could see the tests running?

afrittoli commented:
Thanks for all the updates @jerop .
I don't see the changes from #706 in here?
Thank you 🙏

jerop commented:
> Thanks for all the updates @jerop .
> I don't see the changes from #706 in here?
> Thank you 🙏

just pushed it, thanks :)

afrittoli commented:
> > Thanks for all the updates @jerop .
> > I don't see the changes from #706 in here?
> > Thank you 🙏
> 
> just pushed it, thanks :)

Thank you! 
Unit tests are running for PIP now, it looks like there is a failure.

jerop commented:
> Unit tests are running for PIP now, it looks like there is a failure.

yes, needed to add the event from the permanent error I added, fixed now 

thanks for the reviews @afrittoli!

afrittoli reviewed (commented):
/lgtm
